### PR TITLE
xds: allow tracking the status of DS components

### PR DESCRIPTION
**Pull request by alyssawilk**
As with the last pass I think if we go with this we're going to want to eventually

implement this for all DS classes
add load time as something we track

but let's optionally add one component at at time

Major design questions mostly documented as FIXMES include
Is the proto structure and naming good beyond LDS
How should we handle listeners that never load (config push with new listener that's rejected as broken, config is rolled back and never rolled forward).  I'm inclined to say if a name isn't in a live listener, we set a configurable timer for clean up.

Description: Tracking load status for LDS using the ConfigDump protos
Risk Level: Low: additional proto state and some memory overhead
Testing: NOT DONE YET
Docs Changes: n/a
Release Notes: n/a
#8039

Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

### CODE REVIEW

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 1]
```diff
@@ -45,6 +45,22 @@ message BootstrapConfigDump {
...
+  string name = 1;
+
+  // The timestamp of the last update attempt.
+  google.protobuf.Timestamp last_update = 2;
+
+  // Simple tracking if the last update attempt was successful.
+  bool update_successful = 3;
+
+  // Details about the last update attempt
+  string last_update_details = 4;
```
Comment (ramaraochavali) [inline api/envoy/admin/v2alpha/config_dump.proto]: Based on code, it seems it is only populated on error - probably be explicit about it and call `last_error_details`?

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 2]
```diff
@@ -45,6 +45,22 @@ message BootstrapConfigDump {
   google.protobuf.Timestamp last_updated = 2;
 }
 
+// This is a general purpose proto for dynamic configuration to track information about both
+// currently loaded configuration and the most recent attempt to load configuration.
+message ConfigUpdateState {
+  // The name of the entity whose state is tracked here.
+  string name = 1;
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply api/envoy/admin/v2alpha/config_dump.proto]: So as commented below I don't think that's the right choice here

  // Review: Should this be per DynamicListener?
  // Really per name we want the latest where warming > active > draining and
  // recent loads get mixed in so I think global is correct.

It may be that it makes sense for the other DSs and the listener is weird because it has three versions live, but I think it makes sense to have one state not 4x, so I think it ought to be global 
Basically if we want to know current status you'd have to figure out which of active, warming and draining to look at, and none of those would include "failed to load initially" where I think we want that staus visible, so I think global is the right thing here?

[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/admin/v2alpha/config_dump.proto]: Arguably, from a top-down perspective, we should just have one way of getting at data that belongs to any listener. Can we just have a listener list full stop, and an enum indicating whether warming/active/draining? This is v2alpha, so we can just make the change without "breaking" the API. @mattklein123 thoughts?

[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/admin/v2alpha/config_dump.proto]: Stepping back, what are we trying to achieve here? Basically we just want to know for a named listener what its state is? If so yes I think it's reasonable to rejigger the output so we have a list of listeners, within each listener we have its state, and then if there is a warming/draining secondary config would could optionally print that?

[the remark this answers is not in the document]
Comment (alyssawilk) [reply api/envoy/admin/v2alpha/config_dump.proto]: I want to be able to at a glance for any reloadable component in the system see
1) details of the last successful load (when it happened, time it took, config which we tried to load) and 
2) details iff the last load failed (when it happened, time it took to fail, [eventually] config we tried to load, error message)

So if we have [ successful load ] [ failed load] I'd want to see both, but if we have [failed load] [ successful load] we'd have a green dashboard

I don't mind doing a larger refactor where we track listeners, each listener has a LoadState (which may look like a combination of ConfigUpdateState and DynamicListener) for last_successful_load and last_failed_load, but it still begs the question if a listener has never loaded, how we want to handle that, because I'd still like to have "recent" failures visible on the dashboard even if a listener never successfully loaded.  Harvey, would you suggest the listener list just have {warming / active / draining / inoperational } where again we'd have to reap inoperational by [some configuration values] so we didn't just have a slow leak with stale failures?

**Comment by alyssawilk** [comment]
OK, per discussion on the call, restructured for active/warming/draining inside a general config object, and did examples of config with active/warming/draining and without.

this is super super WIP - let's talk about structure (lots of FIXME discussion points) and then I'll fix typos and comment all the things and do the actual code :-)

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 3]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
     api.v2.Listener listener = 2;
 
     // The timestamp when the Listener was last updated.
+    // FIXME: reviewer we may eventually want last updated and last _successfully_ updated.
+    // Given similarity between last_updated (attempt or success) and
+    // last_update should this maybe be last_update_attempt so last_update
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: Yeah, I would make this `last_update_success` and `last_update_attempt` explicitly now.

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 4]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
...
+    // matches?
     google.protobuf.Timestamp last_updated = 3;
   }
 
+  // Describes a dynamically loaded listener via the LDS API.
+  // FIXME rename? because otherwise v2->alpha is super confusing but this seems
+  // like the right name :-/
+  message DynamicListener {
+    // FIXME worth splitting name out for easy reference, duplicate cruft?
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: Yeah, this is probably reasonable, in particular as this is going to be the basis of JSON dump, so I'm +1 on what you have here.

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 5]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
...
+    // The listener state for any active listener by this name.
+    DynamicListenerState active_state = 2;
+
+    // The listener state for any warming listener by this name.
+    DynamicListenerState warming_state = 2;
+
+    // The listener state for any draining listener by this name.
+    DynamicListenerState draining_state = 3;
+
+    // FIXME alternately we could just have
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: Sub-message is good; it gives room to grow without clutter later.

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 6]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
...
+    // DynamicListenerState failure_state = 4;
+    // string error_details = 5;
+    // rather than the sub-message
+    //
+    // FIXME alternately alternately we could have shared UpdateErrorState with
+    // a Message error_state (which would be DynamicListenerState for listeners,
+    // DynamicClusterState for clusters, and error details for all.
+    // I like shared structures but hate "casting" so I'm scrapping reuse.
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: I would vote on a shared structure. We could put the listener config inside a `google.protobuf.Any` (as long as the JSON dump looks sane, which I think it should as we know the descriptors).

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 7]
```diff
@@ -205,60 +257,76 @@ message ScopedRoutesConfigDump {
     string version_info = 2;
 
     // The scoped route configurations.
-    repeated api.v2.ScopedRouteConfiguration scoped_route_configs = 3;
+    repeated api.v2.ScopedRouteConfiguration scoped_routes = 3;
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: Prefer to avoid gratuitous renames, even within an alpha version, but that's just me :P

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 8]
```diff
@@ -172,30 +205,49 @@ message RoutesConfigDump {
...
   }
 
-  // The statically loaded route configs.
-  repeated StaticRouteConfig static_route_configs = 2;
+  message DynamicRoute {
+    // The name or uniqe id of this route, pulled from config.
+    string name = 1;
+
+    // The route state for any active route by this name.
+    DynamicRouetState route_state = 2;
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: This looks like a type :)

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 9]
```diff
@@ -205,60 +257,76 @@ message ScopedRoutesConfigDump {
...
     string name = 1;
 
-    // This is the per-resource version information.
-    string version_info = 2;
+    // The scoped route state for any active route by this name.
+    DynamicScopedState scoped_route_state = 2;
 
-    // The timestamp when the secret was last updated.
-    google.protobuf.Timestamp last_updated = 3;
+    message RouteUpdateErrorState {
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: Where is this used?

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 10]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
...
     // The timestamp when the Listener was last updated.
+    // FIXME: reviewer we may eventually want last updated and last _successfully_ updated.
+    // Given similarity between last_updated (attempt or success) and
+    // last_update should this maybe be last_update_attempt so last_update
+    // matches?
     google.protobuf.Timestamp last_updated = 3;
   }
 
+  // Describes a dynamically loaded listener via the LDS API.
+  // FIXME rename? because otherwise v2->alpha is super confusing but this seems
```
Comment (mattklein123) [inline api/envoy/admin/v2alpha/config_dump.proto]: Sorry what do you mean here?

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 4 above)
Comment (mattklein123) [reply api/envoy/admin/v2alpha/config_dump.proto]: The other option is to just use a map below which is name -> DynamicListener?

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 6 above)
Comment (mattklein123) [reply api/envoy/admin/v2alpha/config_dump.proto]: I'm fine either way

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 11]
```diff
@@ -70,9 +69,52 @@ message ListenersConfigDump {
...
+    // string error_details = 5;
+    // rather than the sub-message
+    //
+    // FIXME alternately alternately we could have shared UpdateErrorState with
+    // a Message error_state (which would be DynamicListenerState for listeners,
+    // DynamicClusterState for clusters, and error details for all.
+    // I like shared structures but hate "casting" so I'm scrapping reuse.
+    message ListenerUpdateErrorState {
+      // What this listener would have looked like if it had succeeded
+      // FIXME is it confusing to use last_updated to last_attempted_update?
```
Comment (mattklein123) [inline api/envoy/admin/v2alpha/config_dump.proto]: A bit, yes, but if documented it's probably fine, as the time it failed is actually useful IMO.

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 10 above)
Comment (alyssawilk) [reply api/envoy/admin/v2alpha/config_dump.proto]: Sorry it was unclear
Prior to my changes, DynamicListener was a structure which had {list of active, list of warming, list of draining}
With this PR the message DyamicListener is now {?active? ?warming? ?draining? } and we have a list of them.

Basically is it Ok to reuse the name or will it cause more confusion

Comment (mattklein123) [reply api/envoy/admin/v2alpha/config_dump.proto]: I think it's fine to pick the optimal name for the current structure, given this is still an alpha proto.

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 7 above)
Comment (alyssawilk) [reply api/envoy/admin/v2alpha/config_dump.proto]: Does it count as a rename given it's now in a new encapsulating message?
I figure as long as we're moving things around better to have consistency but I'm Ok with changing back too

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 10 above)
Comment (alyssawilk) [reply api/envoy/admin/v2alpha/config_dump.proto]: Cool, that's the feedback I was looking for :-)

OK in general, I moved this to a reusable proto with Any, cleaned up the FIXMES and restructured the rest of the protos.   I think we're in enough agreement on structure I'll start in on code changes tomorrow, but please weigh in with any other API suggestions to minimize that code churn!

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 12]
```diff
@@ -34,14 +34,25 @@ message ConfigDump {
   repeated google.protobuf.Any configs = 1;
 }
 
+message UpdateFailureState {
+  // What the component configuration would have been if the update had succeeded.
+  google.protobuf.Any failed_state = 1;
```
Comment (mattklein123) [inline api/envoy/admin/v2alpha/config_dump.proto]: nit: I would probably call this `failed_configuration` but I don't feel that strongly about it.

File: api/envoy/admin/v2alpha/config_dump.proto (hunk 7 above)
Comment (mattklein123) [reply api/envoy/admin/v2alpha/config_dump.proto]: Per above my preference would be to rename here to the best name since we are doing substantial movement anyway.

**Comment by alyssawilk** [comment]
Ok, I think it's worth leaving this as a reference API PR, and I want to let @htuch look and comment on wednesday, so moving LDS prototype to 8441 with the rename you asked for

File: api/envoy/admin/v2alpha/config_dump.proto [hunk 13]
```diff
@@ -205,30 +240,43 @@ message ScopedRoutesConfigDump {
...
   // The statically loaded scoped route configs.
-  repeated InlineScopedRouteConfigs inline_scoped_route_configs = 1;
+  repeated InlineScopedRoutes inline_scoped_routes = 1;
 
   // The dynamically loaded scoped route configs.
-  repeated DynamicScopedRouteConfigs dynamic_scoped_route_configs = 2;
+  repeated DynamicScopedRoute dynamic_scoped_routes = 2;
+
+  // Set if the last update failed, cleared after the next successful update.
+  UpdateFailureState error_state = 3;
```
Comment (htuch) [inline api/envoy/admin/v2alpha/config_dump.proto]: @stevenzzzz should this be on the `DynamicScopedRoute` granularity? I think we decided this API was basically a CDS-style state-of-the-world, so each individual named entity might have a distinct reason for failed update.

Comment (stevenzzzz) [reply api/envoy/admin/v2alpha/config_dump.proto]: The per-route-configuration update error status is kept by the above DynamicRoute. For SRDS I think this is fine. There is only one error status for a SRDS delta/SotW update, for the SotW form, it's the first whatever error of that update, for the delta-form, it's currently a concatenated error string for all the parsing/validating errors in that update.

Comment (htuch) [reply api/envoy/admin/v2alpha/config_dump.proto]: Well, the same could be said about CDS. But, if you look at what is done there, we break apart the per-resource errors and report them individually. We should do the same for SRDS I think.

Comment (stevenzzzz) [reply api/envoy/admin/v2alpha/config_dump.proto]: I agree we can keep a error status in the per-route-configuration as well.
Currently the error status could be one of "validation/invalid/duplicate" etc, as the scopedRouteConfiguration is simply a pointer to RDS + a key-builder. (it's not that heavy weight as a Cluster).

Also noteworthy to populate per-RouteConfigurationConfig error status, in the SotW for CDS and SRDS alike, would be interesting as they return on the first error.
